### DISCUSSION TITLE
Sync promises

### DISCUSSION THREAD

**Discussion by papb** [mmkal/plv8-git#95] [Ideas]
Hi :grin: 

Your idea of sync-promises caught my attention because I implemented something like that once as well.

I am wondering - do you know if it is possible to spawn a subprocess with `.spawnSync` in plv8 context? Since plv8 seems based on V8, maybe it is possible... And in this case you could use [make-synchronous](https://github.com/sindresorhus/make-synchronous) instead of taking all this SyncPromise hassle.

Another option would be to use [p-syncy](https://github.com/Richienb/p-syncy) which uses deasync under the hood...

By the way, there are comments in your code saying that `.catch` is not supported but I recall managing to make `.catch` work as well.

**Comment by mmkal** [comment]
Hi!

Sadly (or maybe not sadly?) neither of those will work. plv8 doesn't have anything like `child_process`. It's a ["trusted" language](https://www.postgresql.org/docs/8.0/xplang.html#:~:text=The%20languages%20PL%2FpgSQL%2C%20PL,should%20not%20be%20marked%20trusted.), which means it can't really do anything except in memory transformations/maths. No network/filesystem access, and definitely not spawning processes. I believe there's a version of plv8 that's _untrusted_, but I don't see the value in making that a requirement. Spawning subprocesses would be dangerous and almost certainly be bad for performance. It'd also be horrendously complicated to get working, even in "untrusted" mode - which would mean this would no longer work in AWS RDS/Azure postgres etc. All that to save maintaining 44 lines of code 😅

Having said that - I did try to find a package that implemented this already. But most tend to go "the other way". async/await is the shiny new thing (since 2015ish) and I couldn't find any that did this.

About `.catch` - you're right. It'd definitely be possible to support it, by making sync-promise lazy. One approach would be wrapping IOEither from [fp-ts](https://npmjs.com/package/fp-ts). It's basically what sync-promise is aiming to be, but uses static `map`/`chain`/`mapLeft` functions rather than `.then` and `.catch`. I should probably edit the comment saying "catch isn't supported because the callbacks are run synchronously" to "catch isn't supported because I haven't come across any code paths that depend on catching".

**Comment by mmkal** [comment]
Sorry @papb I didn't mean to close this. I'll leave open until there's some sort of support for `.catch`, or I find another dependency that could be used. The dependency would have to be as small as possible though!
